### PR TITLE
Meta constructor for `BlockDiagLazyTensor` to fix its `matmul` in special case

### PR DESCRIPTION
**Pull request by SebastianAment**
`BlockDiagLazyTensor`'s `matmul` method fails with a dimension mismatch, when `BlockDiagLazyTensor` is constructed with a `DiagLazyTensor` as the `base_lazy_tensor` argument. We hit this case in a BoTorch test, and I added it to the `TestBlockDiagLazyTensorMetaClass` below.

@Balandat and I thought that intercepting the construction of the `BlockDiagLazyTensor` and returning a vanilla `DiagLazyTensor` in this case would be a nice solution, instead of introducing another special case to `matmul`. This PR implements this feature via the `_MetaBlockDiagLazyTensor` constructor.

### CODE REVIEW

**Comment by gpleiss** [comment]
Hey @SebastianAment could you re-open this PR in the GitHub.com/cornellius-gp/linear_operator repo? We're about to switch over to using that, and deprecating all of gpytorch's lazy tensor stuff. (Either later today or tomorrow morning, [this PR](https://github.com/cornellius-gp/gpytorch/pull/2027) is finally going to be merged in!)

File: gpytorch/lazy/block_diag_lazy_tensor.py [hunk 1]
```diff
@@ -9,8 +10,26 @@
...
+# BlockDiagLazyTensor(base_lazy_tensor, block_dim=-3) to return a DiagLazyTensor
+# if base_lazy_tensor is a DiagLazyTensor itself
+class _MetaBlockDiagLazyTensor(ABCMeta):
+    def __call__(cls, base_lazy_tensor, block_dim=-3):
+        from .diag_lazy_tensor import DiagLazyTensor
+        if cls is BlockDiagLazyTensor and isinstance(base_lazy_tensor, DiagLazyTensor):
+            if not block_dim == -3:
+                raise NotImplementedError(
+                    f"Passing a base_lazy_tensor of type DiagLazyTensor to the constructor"
+                    "of BlockDiagLazyTensor with block_dim = {block_dim} != -3 is not supported."
```
Comment (Balandat) [inline gpytorch/lazy/block_diag_lazy_tensor.py:23]: Got the f on the wrong line here. Also forgot a space :) 
```suggestion
                    "Passing a base_lazy_tensor of type DiagLazyTensor to the constructor "
                    f"of BlockDiagLazyTensor with block_dim = {block_dim} != -3 is not supported."
```

File: test/lazy/test_block_diag_lazy_tensor.py [hunk 2]
```diff
@@ -67,6 +67,31 @@ def evaluate_lazy_tensor(self, lazy_tensor):
                     actual[i, k, j * 4 : (j + 1) * 4, j * 4 : (j + 1) * 4] = blocks[i, k, j]
         return actual
 
+class TestBlockDiagLazyTensorMetaClass(unittest.TestCase):
+    def test_metaclass_constructor(self):
+        k, n = 3, 5 # number of blocks, block size
```
Comment (Balandat) [inline test/lazy/test_block_diag_lazy_tensor.py:72]: This only tests this without any batch dimension. Could you also add a unit test that checks this for the batched case?

**Comment by SebastianAment** [comment]
Thank you for your feedback! I incorporated it and am moving this PR [over here](https://github.com/cornellius-gp/linear_operator/pull/4).

**Pull request closed without merging**